### PR TITLE
[spirv] Support cbuffer and layout decorations

### PR DESCRIPTION
**Pull request by antiagainst**
* Added support for the cbuffer type
* Added support for accessing elements in cbuffer
* Added support for std140 layout rules

### CODE REVIEW

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -201,6 +203,8 @@ void SPIRVEmitter::doDecl(const Decl *decl) {
     doVarDecl(varDecl);
   } else if (const auto *funcDecl = dyn_cast<FunctionDecl>(decl)) {
     doFunctionDecl(funcDecl);
+  } else if (dyn_cast<HLSLBufferDecl>(decl)) {
+    llvm_unreachable("HLSLBufferDecl should not be handled here");
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: No particular reason. The benefit is that we avoid another `dyn_cast<>`. Not a big deal I think.

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 2]
```diff
@@ -9,23 +9,42 @@
...
-uint32_t TypeTranslator::translateType(QualType type) {
+namespace {
+/// The alignment for 4-component float vectors.
+constexpr uint32_t kStd140Vec4Alignment = 16u;
+
+/// Returns true if the given value is a power of 2.
+inline bool isPow2(int val) { return (val & (val - 1)) == 0; }
+
+/// Rounds the given value up to the given power of 2.
+inline void roundToPow2(uint32_t &val, uint32_t pow2) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Done

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 3]
```diff
@@ -302,6 +327,65 @@ uint32_t TypeTranslator::getComponentVectorType(QualType matrixType) {
...
+    if (isMxNMatrix(fieldType)) {
+      memberAlignment = memberSize = stride = 0;
+      std::tie(memberAlignment, memberSize) =
+          getAlignmentAndSize(fieldType, &stride, isColMajor);
+
+      decorations.push_back(
+          Decoration::getMatrixStride(*spirvContext, stride, index));
+
+      // We need to swap the RowMajor and ColMajor decorations since HLSL
+      // matrices are conceptually row-major while SPIR-V are conceptually
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: Fixed

File: tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.std140.hlsl [hunk 4]
```diff
@@ -49,9 +49,9 @@ cbuffer MyCBuffer {              // Alignment   Offset
 // CHECK-NEXT: OpMemberDecorate %T 1 Offset 16
 // CHECK-NEXT: OpMemberDecorate %T 2 Offset 64
 // CHECK-NEXT: OpMemberDecorate %T 2 MatrixStride 16
-// CHECK-NEXT: OpMemberDecorate %T 2 ColMajor
-// CHECK-NEXT: OpMemberDecorate %T 3 Offset 112
-// CHECK-NEXT: OpMemberDecorate %T 4 Offset 160
+// CHECK-NEXT: OpMemberDecorate %T 2 RowMajor
```
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.std140.hlsl]: wait, shouldn't this be ColMajor?

Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/vk.layout.cbuffer.std140.hlsl]: nvm, got it!

**Pull request merged**